### PR TITLE
Support torch deepseek moe

### PR DESCRIPTION
**Pull request by grimoire**
Add deepseek moe support

require

- [x] #1075 

document would be updated after #1075 merged.

### CODE REVIEW

**Comment by lvhan028** [comment]
```shell
CUDA_VISIBLE_DEVICES=6,7 lmdeploy chat torch /nvme/shared_data/deepseek/deepseek-moe-16b-chat/ --tp 2
```
raises
```
2024-02-26 14:36:44,211 - lmdeploy - ERROR - Init tp model failed with error: [ImportError('This modeling file requires the following packages that were not found in your environment: flash_attn. Run `pip install flash_attn`'), ImportError('This modeling file requires the following packages that were not found in your environment: flash_attn. Run `pip install flash_attn`')]
```
Do we need to install flash-attn?

**Comment by grimoire** [comment]
@lvhan028 yes I guess.

File: lmdeploy/model.py [hunk 1]
```diff
@@ -1013,6 +1013,32 @@ def messages2prompt(self, messages, sequence_start=True):
         return ret
 
 
+@MODELS.register_module(name=['deepseek-chat'])
+class Deepseek(BaseModel):
```
Comment (RunningLeon) [inline lmdeploy/model.py:1017]: @AllentDan Do we need to implement `messages2prompt` method of this class?

File: lmdeploy/model.py [hunk 2]
```diff
@@ -1013,6 +1013,32 @@ def messages2prompt(self, messages, sequence_start=True):
...
+
+    def __init__(self, user='User:', assistant='Assistant:', **kwargs):
+        super().__init__(**kwargs)
+        self._user = user
+        self._assistant = assistant
+        self._bos = '<｜begin▁of▁sentence｜>'
```
Comment (RunningLeon) [inline lmdeploy/model.py:1023]: seems that the eos `<｜end▁of▁sentence｜>` is not used?

File: lmdeploy/model.py (hunk 1 above)
Comment (AllentDan) [reply lmdeploy/model.py:1017]: For api_server's `v1/chat/completions`, yes.

**Comment by lvhan028** [comment]
新模型的支持，PR需要测试以下功能：
1. lmdeploy chat torch（交互式对话）
2. pipeline接口（非交互式）
3. api_server是否正常（非交互式）

File: lmdeploy/model.py (hunk 2 above)
Comment (grimoire) [reply lmdeploy/model.py:1023]: No need

**Comment by grimoire** [comment]
> 新模型的支持，PR需要测试以下功能：
> 
> 1. lmdeploy chat torch（交互式对话）
> 2. pipeline接口（非交互式）
> 3. api_server是否正常（非交互式）

All test passed.

File: lmdeploy/model.py [hunk 3]
```diff
@@ -1013,6 +1013,57 @@ def messages2prompt(self, messages, sequence_start=True):
...
+               session sequence
+        Returns:
+            str: the concatenated prompt
+        """
+        ret = f'{self._user} {prompt}\n\n{self._assistant}'
+        if sequence_start:
+            ret = f'{self._bos}{ret}'
+        return ret
+
+    def messages2prompt(self, messages, sequence_start=True):
```
Comment (RunningLeon) [inline lmdeploy/model.py:1041]: outputs always starts with an empty space, is there something wrong with chat template?
```
Response(text=' Hello! How can I assist you today?', generate_token_len=9, input_token_len=9, session_id=0, finish_reason='stop'), 
Response(text=' Yes, this is a test. Is there anything specific you would like me to test or do for you?', generate_token_len=22, input_token_len=13, session_id=1, finish_reason='stop')
```

File: lmdeploy/model.py [hunk 4]
```diff
@@ -1013,6 +1013,57 @@ def messages2prompt(self, messages, sequence_start=True):
...
+            messages (str | List): user's input prompt
+        Returns:
+            str: the concatenated prompt
+        """
+        if isinstance(messages, str):
+            return self.get_prompt(messages, sequence_start)
+        ret = f'{self._bos}'
+        for _, msg in enumerate(messages):
+            role, content = msg['role'], msg['content']
+            if role == self.user:
```
Comment (AllentDan) [inline lmdeploy/model.py:1056]: no attribute `user`. It is `_user`

File: lmdeploy/model.py [hunk 5]
```diff
@@ -1013,6 +1013,57 @@ def messages2prompt(self, messages, sequence_start=True):
...
+                ret += f'{self._user} {content}\n\n{self._assistant}'
```
Comment (AllentDan) [inline lmdeploy/model.py:1057]: image
Tested OK with this change.

**Review by RunningLeon — APPROVED** [review]
LGTM

File: lmdeploy/model.py [hunk 6]
```diff
@@ -1013,6 +1013,65 @@ def messages2prompt(self, messages, sequence_start=True):
...
+            if role == 'user':
+                ret += f'{self._user} {content}\n\n{self._assistant}'
+                if is_start:
+                    is_start = False
+                else:
+                    ret += f'{self._eos}{ret}'
+            elif role == 'assistant':
+                ret += f' {content}\n\n'
+            else:
+                raise Exception(f'Only have {self.user} and '
```
Comment (RunningLeon) [inline lmdeploy/model.py:1069]: change to `{self.user}` -> `user` and `{self.assistant}` to  `assistant`

**Review by AllentDan — APPROVED** [review]
LGTM

**Pull request merged**